R/desc_utils.R

```diff
@@ -0,0 +1,208 @@
+#' Read DESCRIPTION file and return list of references.
+#'
+#' Returned list is an union between references specified in `"Config/Needs/verdepcheck"` field and
+#' standard references for all other not covered dependencies.
+#' @importFrom pkgdepends pkg_dep_types parse_pkg_ref
+#' @keywords internal
+#' @examplesIf Sys.getenv("R_USER_CACHE_DIR", "") != ""
+#' d <- desc::desc("!new")
+#' d$set_dep("foo", "Import")
+#' d$set_dep("bar", "Suggest")
+#' d$set_list("Config/Needs/verdepcheck", "foo/bar")
+#' d$set_list("Config/Needs/verdepcheck", "foo/baz") # not in package deps - will be skipped
+#' get_refs_from_desc(d)
+get_refs_from_desc <- function(d) {
+  if (.desc_field %nin% d$fields()) {
+    refs <- list()
+  } else {
+    refs <- lapply(get_desc_field_pkgs(d), pkgdepends::parse_pkg_ref)
+  }
+  all_deps <- d$get_deps()
+  all_deps <- all_deps$package[all_deps$type %in% pkgdepends::pkg_dep_types()]
+  missing_refs <- setdiff(
+    setdiff(all_deps, base_pkgs()),
+    map_key_character(refs, "package")
+  )
+  res <- c(
+    refs,
+    lapply(missing_refs, pkgdepends::parse_pkg_ref)
+  )
+  res_idx <- match(all_deps, map_key_character(res, "package"))
+  res_idx <- res_idx[!is.na(res_idx)]
+  res[res_idx]
+}
+
+#' Get the packages from the custom configuration field
+#' @param d (`desc`) DESCRIPTION object from [desc::description]
+#' @return character string
+#' @keywords internal
+get_desc_field_pkgs <- function(d) {
+  if (!d$has_fields(.desc_field)) {
+    return(character(0))
+  }
+  trimws(strsplit(d$get_field(.desc_field), ",")[[1]])
+}
+
+#' Replace Remotes in the `desc` that have been resolved to a GitHub tag or are
+#' in CRAN
+#'
+#' Replaces any existing Remotes entry with the resolved GitHub tag from
+#' `Config/Needs/verdepcheck`.
+#'
+#' @param d (`desc`) DESCRIPTION object
+#' @keywords internal
+#' @examples
+#' # Example that should replace dplyr & tibble on Remotes but not pkgdepends
+#'
+#' d <- desc::desc(
+#'   file = verdepcheck:::local_description(
+#'     list(
+#'       dplyr = "Import",
+#'       tibble = "Import",
+#'       pkgdepends = "Import"
+#'     ),
+#'     remotes = c(
+#'       "tidyverse/dplyr@*release",
+#'       "tidyverse/tibble@*release",
+#'       "r-lib/pkgdepends@*release"
+#'     ),
+#'     need_verdepcheck = c(
+#'       "dplyr",
+#'       "tibble=tidyverse/tibble@v3.2.1"
+#'     )
+#'   )
+#' )
+#' verdepcheck:::desc_remotes_cleanup(d)
+desc_remotes_cleanup <- function(d) {
```

Comment:
@pawelru 

We still need to use `Remotes` for situations when there's a release on GitHub that doesn't exist on CRAN _(and that version is restricted on the `DESCRIPTION` file)_

#### Context:

When a package has a GitHub release that doesn't exist on CRAN and if we remove the Remotes field, then the `ip$solve()` call will try to resolve 2 refs of the same package `cran::<pkg>` and `Config/Needs/verdepcheck::<pkg>` _(and expects those 2 resolutions to succeed)_

* The first will resolve to the CRAN version and the second to GitHub.
* 👉 The issue appears when the version on CRAN doesn't comply with `DESCRIPTION` restriction.
* For a solution, we would need to add even more code for custom resolution _(effectively mocking `Remotes`)_

This can be seen currently with `formatters (>= 0.5.1)` needed for many of the `insightsengineering` packages _(see example below)_.

It works without the `Remotes` field for the `max` strategy because there's a valid CRAN/PPM release that satisfies the `DESCRIPTION`, it doesn't on edge cases like the `formatters`

```R
test_that("remotes is needed when release is on github, but not on CRAN", {
  path_desc <- local_description(
    pkg_list = list("formatters (>= 0.5.1)" = "Imports"),
    need_verdepcheck = "insightsengineering/formatters"
  )

  result <- min_isolated_deps_check(path_desc)

  expect_equal(result$ip$get_solution()$status, "FAILED")

  path_desc_remotes <- local_description(
    pkg_list = list("formatters (>= 0.5.1)" = "Imports"),
    need_verdepcheck = "insightsengineering/formatters",
    remotes = list("insightsengineering/formatters@*release")
  )

  results_remotes <- min_isolated_deps_check(path_desc_remotes)

  expect_equal(results_remotes$ip$get_solution()$status, "OK")
})
```

Reply:
Now that `Remotes` has been removed from insightsengineering packages `verdepcheck` will fail on those packages.

#### What can be done to workaround this?

Whenever a package resolves to a `github_remote_ref` on the `Config/Needs/verdepcheck` we can artificially add that reference to `Remotes`. 

For example with `min_isolated` strategy, `teal.data` is currently failing on verdepcheck:

![image](https://github.com/insightsengineering/verdepcheck/assets/211358/2656867a-c581-498b-8d5a-fb2ba215d338)

It resolves the `Config/Needs/verdepcheck` entry for `teal.logger` to be `teal.logger=insightsengineering/teal.logger@v0.1.1`

We can look to cases when it doesn't resolve to CRAN and add those to remotes.

![image](https://github.com/insightsengineering/verdepcheck/assets/211358/b301d2bd-8e16-47a2-a26a-83959ce3fdc6)

Reply:
I have pushed a commit that aims to address this issue by keeping config/needs/verdepcheck row only so that we would be having only one (successful) resolution. I have checked this locally using your branches and it's all good. Let's try to test it more.

Reply:
The tests I've done with the current `branch` look good. _(current branch as in: that artificially populates `Remotes` with all `verdepcheck` github dependencies)_... The test suite is quite extensive and caught a few issues.

When attempting not to use `Remotes` at all _(with a change on `desc_remotes_cleanup()`)_ it hits a couple of problems

We go back to how `{pkgdepends}` doesn't reconcile packages dependency and version with `Config/Needs/verdepcheck`.

<details>
<summary>⊞ How to test? (click here to expand)</summary>

1. Apply diff below
2. run locally against [`teal.slice@verdepcheck_action`](https://github.com/insightsengineering/teal.slice/tree/verdepcheck_action)

```diff

diff --git a/R/desc_utils.R b/R/desc_utils.R
index 924f1a7..67340d3 100644
--- a/R/desc_utils.R
+++ b/R/desc_utils.R
@@ -94,7 +94,7 @@ desc_remotes_cleanup <- function(d) {
 
   # Create new list of references that will be used as "Remotes"
   new_remotes <- c(
-    map_key_character(desc_field_refs[desc_field_include_ix], "ref"),
+    # map_key_character(desc_field_refs[desc_field_include_ix], "ref"),
     map_key_character(remotes_refs[remotes_include_ix], "ref")
   )
 ```
</details>

#### Q: How flexible are you on populating `Remotes` with `Config/Needs/verdepcheck` versions?

As it is right now, `Remotes` will only be populated with the resolved `Config/Needs/verdepcheck` tags

- for `max` strategy it will be the GH repo name
- for all others it will be specific tags from GH repo
  - It won't use `org/repo@*release` (it will overwrite it if there's a corresponding `Config/Needs/verdepcheck`